### PR TITLE
Reuse bool box objects in UncommonField<bool>.SetValue

### PR DESCRIPTION
**Pull request by stephentoub**
## Description

In some scenarios, these UncommonField<bool>s end up being set very frequently, and each time the value being passed in gets boxed.  We can avoid that by caching and reusing a box for each of the only two possible values (though it looks like the default value is almost always false and so it mainly only boxes true).

## Customer Impact

Unnecessary allocation

## Regression

No

## Testing

CI

## Risk

Minimal.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/wpf/pull/6528)

### CODE REVIEW

File: src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/UncommonField.cs [hunk 1]
```diff
@@ -136,6 +146,11 @@ internal int GlobalIndex
         private int _globalIndex;
         private bool _hasBeenSet;
 
+        /// <summary>A lazily boxed false value.</summary>
+        private static object s_false;
```
Comment (ThomasGoulet73) [inline src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/UncommonField.cs:150]: You could use [MS.Internal.KnownBoxes.BooleanBoxes](https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/WindowsBase/MS/Internal/KnownBoxes.cs#L11) which is already used in a bunch of places in this repo.

Comment (stephentoub) [reply src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/UncommonField.cs]: Wasn't aware that existed.  Even better.

File: src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/UncommonField.cs [hunk 2]
```diff
@@ -54,12 +53,24 @@ public void SetValue(DependencyObject instance, T value)
                 // Set the value if it's not the default, otherwise remove the value.
                 if (!EqualityComparer<T>.Default.Equals(value, _defaultValue))
                 {
-                    instance.SetEffectiveValue(entryIndex, null /* dp */, _globalIndex, null /* metadata */, value, BaseValueSourceInternal.Local);
+                    object valueObject;
+
+                    if (typeof(T) == typeof(bool))
```
Comment (lindexi) [inline src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/UncommonField.cs:58]: May I know the performance difference between `typeof(T) == typeof(bool)` and `T is bool`? Thank you.

Comment (stephentoub) [reply src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/UncommonField.cs:58]: `typeof(T) == typeof(bool)` and `value is bool` should both be optimized to a const value by the JIT in a release build and tier 1 compilation.  In tier 0, it's possible `value is bool` could box.

Comment (lindexi) [reply src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/UncommonField.cs:58]: Thank you very much.

**Review by dipeshmsft — APPROVED** [review]

**Pull request merged**